### ISSUE TITLE
Conflicting PreserveAttribute in RefitStubs.g.cs when using InternalsVisibleTo attribute

### ISSUE THREAD

**Issue by onarres** [reactiveui/refit#565]
I have two projects in a solution, my main assembly named Foo and a unit test assembly named Foo.Tests.
Both assemblies are using Refit for my particular testing needs. This was working perfect until I needed to add unit tests that needed access to an internal class in Foo. I added the InternalsVisibleTo attribute to the AssemblyInfo.cs in Foo so Foo.Tests had access.  Doing this causes this error:

`obj/Release/netcoreapp2.0/RefitStubs.g.cs(43,6): error CS0436: The type 'PreserveAttribute' in '/home/build/Tests/Foo.Tests/obj/Release/netcoreapp2.0/RefitStubs.g.cs' conflicts with the imported type 'PreserveAttribute' in 'Foo, Version=0.0.0.999, Culture=neutral, PublicKeyToken=null'. Using the type defined in '/home/build/Tests/Foo.Tests/obj/Release/netcoreapp2.0/RefitStubs.g.cs'. [/home/build/Tests/Foo.Tests/Foo.Tests.csproj]`

The problem is that when refit generates it stubs, it uses the same namespace and/or class names in both projects and then the compiler throws a warning because it doesn't know which one to use. Warnings at my company are treated as failures so I needed to fix this issue.

My current work around was to remove Refit from Foo.Tests and add all my test classes directly in Foo. Since Refit was not added to Foo.Tests, I don't have the problem.  Obviously, this is not ideal. 

One proposed solution is the ability to change the namespace of the stubs so there is not a conflict in this specific use case.

**Comment by onovotny** [comment]
He’s using InternalsVisibleTo and also using refit in the unit test class. 

The fix would be to add a mangled assembly name to the generated namespace.

**Comment by Styxxy** [comment]
@onovotny Might it be a suggestion to use the project `$(RootNamespace)`? Maybe even add a possibilty to overrule this in the project file?

Eg:

**refit.targets**
```xml
<Project>
  <!-- ... -->

  <PropertyGroup>
    <RefitInternalNamespace Condition=" '$(RefitInternalNamespace)' == '' ">$(RootNamespace)</RefitInternalNamespace>
  </PropertyGroup>

  <!-- ... -->

  <Target Name="GenerateRefitStubs" BeforeTargets="CoreCompile">

    <!-- ... -->
    
    <GenerateStubsTask
        SourceFiles="@(Compile)"
        BaseDirectory="$(MSBuildProjectDirectory)" 
        OutputFile="$(IntermediateOutputPath)\RefitStubs.g.cs"
        RefitInternalNamespace="$(RefitInternalNamespace)"
        />

    <!-- ... -->    
  </Target>
```

We pass this _RefitInternalNamespace_ along to the InterfaceStubGenerator class, which it will include when generating the mustache template.

Any thoughts on this approach?

**Comment by onovotny** [comment]
The `RootNamespace` is more likely to conflict again if it's repeated. What if we just use `AssemblyName` since you can't have two different assemblies with the same name in the output directory?

**Comment by Styxxy** [comment]
Fixed in version 4.6.58 on Nuget. You can update to that version ;-).
